### PR TITLE
Hierarchical forecasting notebook

### PR DESCRIPTION
**Pull request by ivanmkc**
This is a hierarchical forecasting notebook demonstrating the use of the new hierarchical forecasting parameters for AutoML Forecasting.

If you are opening a PR for `Official Notebooks` under the [notebooks/official](https://github.com/GoogleCloudPlatform/vertex-ai-samples/tree/main/notebooks/official) folder, follow this mandatory checklist:
- [x] Use the [notebook template](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/notebooks/notebook_template.ipynb) as a starting point.
- [x] Follow the style and grammar rules outlined in the above notebook template.
- [ ] Verify the notebook runs successfully in Colab since the automated tests cannot guarantee this even when it passes.
- [x] Passes all the required automated checks. You can locally test for formatting and linting with these [instructions](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/CONTRIBUTING.md#code-quality-checks).
- [ ] You have consulted with a tech writer to see if tech writer review is necessary. If so, the notebook has been reviewed by a tech writer, and they have approved it.
- [x] This notebook has been added to the [CODEOWNERS](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/notebooks/official/CODEOWNERS) file under the `Official Notebooks` section, pointing to the author or the author's team.
- [x] The Jupyter notebook cleans up any artifacts it has created (datasets, ML models, endpoints, etc) so as not to eat up unnecessary resources.

### CODE REVIEW

**Comment by ivanmkc** [comment]
go/gcbrun

**Comment by andrewferlitsch** [comment]
running test > 4 days
Apparently Cloud build just fixed that bug in the latest release
Might have to gcbrun again to see the fix

**Comment by reznitskii** [comment]
<p>Forecasting optimization objectives are given here:</p><p>https://cloud.google.com/vertex-ai/docs/tabular-data/forecasting/train-model#optimization-objectives</p><p> </p><p>Looks like all three of the objectives you have here are valid. </p><p> </p><p>Why does the text say AutoMLTabularTrainingJob but the code says AutoMLForecastingTrainingJob? I think the text needs to be cleaned up, variable optimization_prediction_type is not set in the code, variable column_transformations is called column_specs in the code. </p>

---

 <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/664/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by reznitskii** [comment]
<p>If I understand correctly, the rolling window feature is distinct and separate from hierarchical forecasting. We could add a demonstration of it to the primary forecasting notebook: https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/notebooks/official/automl/sdk_automl_tabular_forecasting_batch.ipynb</p>

---

 <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/664/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by reznitskii** [comment]
<p>Is it possible to demonstrate the full set of hierarchical forecasting features within a single notebook? If not and we choose to demonstrate just <code>group_temporal_total_weight,</code>I think we should be clear about this from the beginning, in the tutorial introduction.</p>

---

 <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/664/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by reznitskii** [comment]
<p>1 - Have you talked to Andrew Leach?</p><p> </p><p>2 - Yes, it would be good to compare the results of the hierarchical and the non-hierarchical model, but can we just load the result of the non-hierarchical model for this dataset from somewhere? I would prefer to avoid creating / training the the non-hierarchical model here. </p>

---

 <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/664/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by reznitskii** [comment]
<p>Is it possible to include a picture that demonstrates the hierarchy we use here?</p><p>Is "product_type" a subset of "product_category", which is a subset of "product"?</p>

---

 <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/664/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by ivanmkc** [comment]
<p>Just for the user to understand what the dataset looks like.</p>

---

 <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/664/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by ivanmkc** [comment]
<p>Are they mutually exclusive?</p>

---

 <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/664/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by ivanmkc** [comment]
<p>AFAIK, it must be empty for the prediction to work.</p>

---

 <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/664/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by AndrewLeach** [comment]
<p>Early stopping is not supported on Vertex Forecast yet, it is in the H2 roadmap though.</p>

---

 <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/664/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by AndrewLeach** [comment]
<p>What hierarchical configuration to use depends on what the goal is. Are there product category level metrics we are inspecting (group weight with product category as group column)? Are there season total metrics we are inspecting (temporal weight with no group columns)? Are we checking the overall bias (group weight with no group columns)?</p>

---

 <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/664/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by ivanmkc** [comment]
<p>Got it, thanks!</p>

---

 <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/664/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by ivanmkc** [comment]
<p>Removed.</p>

---

 <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/664/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

File: notebooks/official/automl/sdk_automl_forecasting_hierarchical_batch.ipynb [hunk 1]
```diff
@@ -0,0 +1,1525 @@
+{
```
Comment (ivanmkc) [inline notebooks/official/automl/sdk_automl_forecasting_hierarchical_batch.ipynb:1]: <p>@reznitskii, I added some text that Andrew Leach and I put together for the hierarchical parameters.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/664/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='51'/>

**Comment by ivanmkc** [comment]
<p>Added</p>

---

 <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/664/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by ivanmkc** [comment]
<p>Adding... One second.</p>

---

 <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/664/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by ivanmkc** [comment]
<p>Done.</p>

---

 <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/664/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by reznitskii** [comment]
<p>We could say: "We are predicting the value of 'sales'." Note that this column is empty in the dataset used for prediction."</p>

---

 <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/664/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by reznitskii** [comment]
<p>Any updates here?</p>

---

 <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/664/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by reznitskii** [comment]
<p>Any updates here?</p>

---

 <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/664/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by reznitskii** [comment]
<ol><li>Any update here?</li><li>I'm not sure what you mean. Can we assume that the dataset is going to be fixed and the non-hierarchical model training algorithm will be fixed?</li></ol><p> </p><p>I think comparing with the ground truth is not so meaningful here, the point is to justify doing model training with the additional hierarchical component. </p>

---

 <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/664/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by ivanmkc** [comment]
<p>Coming back to this. I will move this to the beginning as suggested.</p>

---

 <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/664/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by ivanmkc** [comment]
<p>Will do.</p>

---

 <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/664/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by ivanmkc** [comment]
<p>No response.</p>

---

 <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/664/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by ivanmkc** [comment]
<p>No updates from Andrew. </p><p> </p><p>I am waiting on an updated dataset to get a better prediction result.</p>

---

 <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/664/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by ivanmkc** [comment]
<p>Will fix</p>

---

 <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/664/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by ivanmkc** [comment]
<p>Will do.</p>

---

 <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/664/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by ivanmkc** [comment]
<p>Good point. I will remove the ground truth comparison.</p>

---

 <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/664/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by ivanmkc** [comment]
<p>Done.</p>

---

 <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/664/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by ivanmkc** [comment]
<p>Done.</p>

---

 <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/664/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Review by soheilazangeneh — APPROVED** [review]
LGTM! Only left few minor comments!

**Comment by ivanmkc** [comment]
<p>I think the comparison is out of scope as we move towards notebooks that focus on showcasing how to use the API.</p>

---

 <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/664/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by ivanmkc** [comment]
<p>Added more details.</p>

---

 <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/664/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Pull request merged**